src/nvim/os/keyboard.c

```diff
@@ -0,0 +1,222 @@
+    return false;
+  }
+
+  _XkbStateRec xkb_state;
+  const _XStatus status = _XkbGetState(display, _XkbUseCoreKbd, &xkb_state);
+  if (status) {
+    *err = "Non-zero status returned by XkbGetState";
+    CLOSE_DISPLAY(display);
+  }
+  *mods = xkb_state.data[kLockedModsOffset];
+
+  CLOSE_DISPLAY(display);
+  return true;
+#undef CLOSE_DISPLAY
+}
+
+#elif defined(WIN32)
+/// Get the status of the Capslock, Numlock and Scrolllock in windows
+bool os_get_locks_status(char *const mods, const char **const err)
+{
+  *mods = 0;
+  *err = NULL;
+  return (GetKeyState(VK_CAPITAL) & 0x0001) != 0;
+  if (GetKeyState(VK_CAPITAL) & 0x0001) {
+    *mods = *mods | kCapsLock;
+  }
+  if (GetKeyState(VK_SCROLL) & 0x0001) {
+    *mods = *mods | kScrollLock;
+  }
+  if (GetKeyState(VK_NUMLOCK) & 0x0001) {
+    *mods = *mods | kNumLock;
+  }
+  return true;
+}
+
+#elif defined(__APPLE__)
+/// Get the status of the Capslock, Numlock and Scrolllock in MacOS
+static bool os_get_locks_status(char *const mods, const char **const err)
+{
+  *err = NULL;
+  io_service_t ios;
+  io_connect_t ioc;
+  CFMutableDictionaryRef mdict;
+  bool stateCaps, stateNums, stateScroll;
+  mdict = IOServiceMatching(kIOHIDSystemClass);
+  ios = IOServiceGetMatchingService(kIOMasterPortDefault,
+                                    (CFDictionaryRef)mdict);
+  if (!ios) {
+    if (mdict) {
+      CFRelease(mdict);
+    }
+    *err = "Unable to start service to query data";
+    return false;
+  }
+
+  if (IOServiceOpen(ios, mach_task_self(), kIOHIDParamConnectType,
+                    &ioc) != kIOReturnSuccess) {
+    IOObjectRelease(ios);
+    *err = "Unable to query data";
+    return false;
+  }
+  IOObjectRelease(ios);
+
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateCaps) != kIOReturnSuccess) {
+    *err = "Unable to lock the state";
+    return false;
+  }
+
+  if (IOHIDGetModifierLockState(ioc, kIOHIDCapsLockState,
+                                &stateNums) != kIOReturnSuccess) {
+    *err = "Unable to lock the state";
+    return false;
+  }
+  if (stateCaps) {
+    *mods = *mods | kCapsLock;
+  }
+  if (stateNums) {
+    *mods = *mods | kNumLock;
+  }
+  IOServiceClose(ioc);
+  return true;
+}
+#endif
+
+/// A helper function to be used by `nvim_get_keyboard_mods_state` API function
+/// giving the information about the Capslock, Numlock and Scrollock state.
+///
+/// @param dict[out]  The dictionary holding the information of the
+///                   different mods
+///
+/// @param err[out]   The error if encountered, otherwise will be NULL
+///
+/// @return           `true` if no error occurs
+///                   otherwise `false`
+bool os_mods_status(Dictionary *const dict, const char **const err )
+{
+  char mods = 0;
+  *err = NULL;
+  bool status = os_get_locks_status(&mods, err);
+  if (!status) {
+      return false;
+  }
+  if (mods & kCapsLock) {
+    PUT(*dict, "Capslocks", BOOLEAN_OBJ(true));
+  }
```

Comment:
I would rather do `PUT(*dict, "capslock", BOOLEAN_OBJ(mods & kCapsLock));`: it is almost always easier to just write `ret["capslock"]` then any kind of checking whether key is in dictionary, at best it would be something like moving from `d[k]` to `d.get(k, default)`. Though additionally making keys also present allows distinguishing situation between “this kind of lock is checked” vs “this kind of lock is not checked at all”. And note that dictionary keys are inconsistent here (plural vs singular) and it is better to not have capital letter.

Also given your comment about OS X I should say that if scroll lock is not checked there then it should not appear in dictionary at all exactly for the purpose of telling the caller that he may not get knowledge of scroll lock status.

Reply:
Yes, I also find it right now